Домашка спринт №23

AlekseyChist opened the pull request.

AlekseyChist committed:
Домашка Спринт 23
Шаг 1. Иконка приложения

AlekseyChist committed:
Домашка Спринт 23
Шаг 2. Экран «Плейлист»

Нужно привести к стилю макета

AlekseyChist committed:
Домашка Спринт 23
Шаг 2. Экран «Плейлист»

Привел к стилю макета

AlekseyChist committed:
Домашка Спринт 23
Шаг 3. Список треков в плейлисте

AlekseyChist committed:
Домашка Спринт 23
Шаг 4. DAO для удаления плейлистов и расширяем репозиторий и интерактор

AlekseyChist committed:
Домашка Спринт 23
Шаг 4. Расширил интерактор

AlekseyChist committed:
Домашка Спринт 23
Шаг 4. Обновил viewmodel создал интерактор для функции "Поделиться"

AlekseyChist committed:
Домашка Спринт 23
Шаг 4. Доделал

AlekseyChist committed:
Домашка Спринт 23
Шаг 5. ViewModel для редактирования

AlekseyChist committed:
Домашка Спринт 23
Шаг 5. починил, нужно привести в соответсвие с макетом

AlekseyChist committed:
Домашка Спринт 23
Шаг 5. обновил по цветам

AlekseyChist committed:
Домашка Спринт 23
Шаг 5. поправил кнопки

AlekseyChist committed:
Домашка Спринт 23
Шаг 5. поправил цвета день/ночь

AlekseyChist committed:
Домашка Спринт 23
Шаг 5. причесал файлы

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/domain/usecase/PlaylistInteractor.kt:
```diff
@@ -48,4 +53,54 @@ class PlaylistInteractorImpl(
+        }
+
+        // Количество треков
+        val tracksCountText = when {
```
Воспользуйся лучше plurals, в андроиде уже реализована логика, с помощью которой для каждой локали будет подбираться соответствующее окончание. Почитать о том, как это использовать, можно в официальной документации - https://developer.android.com/guide/topics/resources/string-resource#Plurals

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/adapter/PlaylistTrackAdapter.kt:
```diff
@@ -0,0 +1,71 @@
+                .load(track.artworkUrl100)
+                .placeholder(R.drawable.placeholder_image)
+                .error(R.drawable.placeholder_image)
+                .transform(RoundedCorners(8))
```
В конструктор `RoundedCorners` нужно передавать кол-во пикселей. Это значит, что предварительно число, передаваемое в конструктор, нужно сконвертировать в пиксели, если мы хотим добиться одинаково скругления на девайсах с разной плотностью экрана

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/fragment/EditPlaylistFragment.kt:
```diff
@@ -0,0 +1,93 @@
+            )
+
+            if (coverFile.exists()) {
+                binding.addImageIcon.visibility = View.GONE
```
У `view` есть удобный экстеншен: `View.isVisible = true/false`, которая заменяет привычное нам `View.visibility = View.VISIBLE/View.GONE`

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/fragment/PlaylistDetailFragment.kt:
```diff
@@ -0,0 +1,327 @@
+
+    private fun showDeleteTrackDialog(track: Track) {
+        MaterialAlertDialogBuilder(requireContext())
+            .setMessage("Хотите удалить трек?")
```
Все строки лучше хранить в ресурсах

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/fragment/PlaylistDetailFragment.kt:
```diff
@@ -0,0 +1,327 @@
+                    .load(coverFile)
+                    .placeholder(R.drawable.placeholder_image)
+                    .error(R.drawable.placeholder_image)
+                    .transform(CenterCrop(), RoundedCorners(8))
```
В конструктор `RoundedCorners` нужно передавать кол-во пикселей. Это значит, что предварительно число, передаваемое в конструктор, нужно сконвертировать в пиксели, если мы хотим добиться одинаково скругления на девайсах с разной плотностью экрана

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/viewmodel/EditPlaylistViewModel.kt:
```diff
@@ -0,0 +1,79 @@
+    private var hasNewCover = false
+
+    private val _playlistData = MutableLiveData<Playlist>()
+    val playlistData: LiveData<Playlist> = _playlistData
```
при работе в контексте вьюмодель-экран (активити/фрагмент) - лучше работать с отдельной ui-моделью, нам не требуются все поля из доменной модели + так прощу будет рефакторить отдельные слои не затрагивая большую часть логики

здесь так же стоит продумать общий стейт вместо отдельных лайвдат

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/viewmodel/EditPlaylistViewModel.kt:
```diff
@@ -0,0 +1,79 @@
+            try {
+                // Если выбрана новая обложка, сохраняем её
+                val finalCoverPath = if (hasNewCover && playlistCoverUri != null) {
+                    playlistInteractor.createPlaylist(
```
интерактор создаст отдельный плейлист и мы будем видеть его на ui

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/viewmodel/PlaylistDetailViewModel.kt:
```diff
@@ -0,0 +1,107 @@
+                playlistInteractor.removeTrackFromPlaylist(currentPlaylistId, track.trackId)
+                loadPlaylist(currentPlaylistId)
+            } catch (e: Exception) {
+                // Обработка ошибки
```
можно обойтись без использования try/catch в корутинах - для этого существуют отдельные exceptionHandler'ы:
https://kotlinlang.org/docs/exception-handling.html#coroutineexceptionhandler

sukhoikms27 commented on app/src/main/java/com/example/playlistmaker/media/ui/viewmodel/PlaylistDetailViewModel.kt:
```diff
@@ -0,0 +1,107 @@
+        return "$totalMinutes минут"
+    }
+
+    fun formatTracksCount(count: Int): String {
```
это уже второй метод, где мы делаем разбивку по количеству треков - лучше точно использовать plurals

sukhoikms27 reviewed (commented):
1 - Не соответствие экрана плейлиста макетам (размер кнопок шаринга и трех точек)

<img width="1080" height="2400" alt="Image" src="https://github.com/user-attachments/assets/4c03b1a4-0022-4049-9b6b-e85819706fee" />
макет:

<img width="387" height="855" alt="Image" src="https://github.com/user-attachments/assets/073a8611-8ffb-47be-81f8-80e7412095f8" />
3 - Некорректный текст в контекстном меню:

<img width="1080" height="2400" alt="Image" src="https://github.com/user-attachments/assets/f46b785b-78df-48c0-9170-efeec86a1456" />

6 - При редактировании плейлиста не должно быть нав панели:

<img width="1080" height="2400" alt="Image" src="https://github.com/user-attachments/assets/6f798d3c-c023-4c0b-93f3-6fdb0049fc2f" />

AlekseyChist committed:
Домашка Спринт 23
Исправил:
- Проблемау с редактированием обложки плейлиста
- Добавил метод для сохранения обложки
- Сортировку треков в обратном порядке
- Кнопки
- Текст в меню
- Убрал навигационную панель при редактировании
- Добавил Plurals

AlekseyChist committed:
Домашка Спринт 23
Исправил:
- добавил Extensions.kt
- подправил код

AlekseyChist committed:
Домашка Спринт 23
Исправил:
- размер кнопок

AlekseyChist committed:
Домашка Спринт 23
Исправил:
- починил багу
